--- mz-0.40.orig/src/mops_tcp.c	2010-03-07 07:43:55.000000000 -0500
+++ mz-0.40/src/mops_tcp.c	2011-02-06 22:24:13.000000000 -0500
@@ -33,9 +33,9 @@
 	if (!t) return 1; // delta set to zero means no range
    
 	if (a<b) // regular case
-		result = ceill ((b-a)/t);
+		result = ceil ((b-a)/t);
 	else     // range wraps around
-		result = ceill (((0xffffffff-a) + b)/t);
+		result = ceil (((0xffffffff-a) + b)/t);
    
 	return result;
 }
@@ -53,9 +53,9 @@
 	if (!t) return 1; // delta set to zero means no range
    
 	if (a<b) // regular case
-		result = ceill ((b-a)/t);
+		result = ceil ((b-a)/t);
 	else     // range wraps around
-		result = ceill (((0xffffffff-a) + b)/t);
+		result = ceil (((0xffffffff-a) + b)/t);
    
 	return result;
 }
