--- mz-0.40.orig/CMakeLists.txt	2010-07-07 15:37:05.000000000 -0400
+++ mz-0.40/CMakeLists.txt	2011-02-06 22:43:56.000000000 -0500
@@ -1,11 +1,12 @@
 project(mausezahn)
 
 cmake_minimum_required(VERSION 2.4.0 FATAL_ERROR)
+
 if(COMMAND cmake_policy)
 	cmake_policy(SET CMP0003 NEW)
 endif(COMMAND cmake_policy)
 
-SET(CMAKE_C_FLAGS "-Wall -pipe -fexceptions -fstack-protector --param=ssp-buffer-size=4 -fasynchronous-unwind-tables")
+SET(CMAKE_C_FLAGS "-Wall -pipe -D_FORTIFY_SOURCE=2 -fexceptions --param=ssp-buffer-size=4 -fasynchronous-unwind-tables")
 
 ADD_CUSTOM_TARGET(uninstall
 	COMMAND ${CMAKE_COMMAND} -E echo Use 'xargs rm < install_manifest.txt' to uninstall this program
@@ -13,8 +14,19 @@
 
 include_directories(./ ../ ${CMAKE_CURRENT_BINARY_DIR})
 
+# OpenWRT
+SET(IN_OPENWRT $ENV{IN_OPENWRT})
+IF (IN_OPENWRT)
+       ADD_DEFINITIONS("$ENV{TARGET_LDFLAGS}" "$ENV{TARGET_CPPFLAGS}" "$ENV{TARGET_CFLAGS}")
+       INCLUDE_DIRECTORIES("$ENV{TARGET_LDFLAGS}" "$ENV{TARGET_CPPFLAGS}" "$ENV{TARGET_CFLAGS}")
+ENDIF(IN_OPENWRT)
+
+SET (DEVROOT "$ENV{STAGING_DIR}")
+INCLUDE_DIRECTORIES ("$ENV{STAGING_DIR}/usr/include")
+LINK_DIRECTORIES ("$ENV{STAGING_DIR}/usr/lib")
+INCLUDE_DIRECTORIES( "$ENV{STAGING_DIR}/usr/lib/libnet-1.2.x/include" )
+SET(NET_LIBRARY $ENV{STAGING_DIR}/usr/lib/libnet-1.2.x/lib/libnet.so)
+
 add_subdirectory (src)
 add_subdirectory (doc)
 
-
-
