--- ftx-prog/ftx_prog.c.old	2016-08-01 17:28:07.000000000 +0200
+++ ftx-prog/ftx_prog.c	2016-08-01 21:16:20.626595514 +0200
@@ -1074,7 +1074,7 @@ int main (int argc, char *argv[])
 
 	memset(&ee, 0, sizeof(ee));
 	ee.old_vid = 0x0403;	/* default; override with --old_vid arg */
-	ee.old_pid = 0x6015;	/* default; override with --old_pid arg */
+	ee.old_pid = 0x6014;	/* default; override with --old_pid arg */
 	process_args(argc, argv, &ee);	/* handle --help and --old-* args */
 
 	if (ftdi_usb_open_desc(&ftdi, ee.old_vid, ee.old_pid, NULL, ee.old_serno)) {
